### PR TITLE
Use the placement API

### PR DESCRIPTION
**Pull request by dhaiducek**
Addresses:
- https://github.com/open-cluster-management/backlog/issues/16319

Leaves existing placement keys and adds `placementPath` and `labelSelector` to use the Placement API.

Also, if no placement is given it defaults to using the Placement kind. I think this is fine since the generator would probably update the PlacementBindings for the new Placement and we can note to customers that they'll need to delete leftover PlacementRules on the cluster post-migration. (I'd need to test that...)

I also moved the policy name validation up and updated the error messages after so that we could reference the policy name in the message. That seems helpful to me, but I can revert them if it doesn't make sense.

### CODE REVIEW

File: internal/plugin.go [hunk 1]
```diff
@@ -281,9 +289,19 @@ func (p *Plugin) applyDefaults(unmarshaledConfig map[string]interface{}) {
 			policy.ConsolidateManifests = p.PolicyDefaults.ConsolidateManifests
 		}
 
-		// If both cluster selectors and placement rule path aren't set, then use the
-		// defaults with a priority on placement rule path.
-		if len(policy.Placement.ClusterSelectors) == 0 && policy.Placement.PlacementRulePath == "" {
+		// If both cluster label selectors and placement path aren't set, then use the defaults with a
```
Comment (mprahl) [inline internal/plugin.go:294]: Could use additional variables to make this easier to read?

For example:

```go
plDefaultSet := len(p.PolicyDefaults.Placement.LabelSelector) != 0 || p.PolicyDefaults.Placement.PlacementPath != ""
plrDefaultSet := len(p.PolicyDefaults.Placement.ClusterSelectors) != 0 || p.PolicyDefaults.Placement.PlacementRulePath != ""
```

Then your if statements can be:
```go
if len(policy.Placement.LabelSelector) == 0 && policy.Placement.PlacementPath == "" && plDefaultSet {
```

and

```go
else if len(policy.Placement.ClusterSelectors) == 0 && policy.Placement.PlacementRulePath == "" && plrDefaultSet {
```

File: internal/plugin.go [hunk 2]
```diff
@@ -365,8 +423,38 @@ func (p *Plugin) assertValidConfig() error {
...
+			_, err := os.Stat(policy.Placement.PlacementPath)
+			if err != nil {
+				return fmt.Errorf(
+					"could not read the placement rule path %s",
+					policy.Placement.PlacementPath,
+				)
+			}
+		}
 
 		seen[policy.Name] = true
```
Comment (mprahl) [inline internal/plugin.go]: Could you move this up to right after `if seen[policy.Name] {`? Then it's easier to read since where it is updated and used are in the same section of code.

File: internal/plugin.go [hunk 3]
```diff
@@ -365,8 +423,38 @@ func (p *Plugin) assertValidConfig() error {
...
+					"policy '%s' may not use both Placement and PlacementRule kinds", policy.Name,
+				)
+			}
+		}
+	}
+
+	// Validate only one type of placement kind is in use
+	if p.plrCount.pl != 0 && p.plrCount.plr != 0 {
+		return fmt.Errorf(
+			"may not use a mix of Placement and PlacementRule for policies--found %d Placement and %d PlacementRule",
```
Comment (mprahl) [inline internal/plugin.go]:
```suggestion
			"may not use a mix of Placement and PlacementRule for policies; found %d Placement and %d PlacementRule",
```

File: internal/plugin.go [hunk 4]
```diff
@@ -424,10 +512,38 @@ func (p *Plugin) getPlrFromPath(plrPath string) (string, map[string]interface{},
 	var name string
 	var rule map[string]interface{}
 	for _, manifest := range *manifests {
-		if kind, _, _ := unstructured.NestedString(manifest, "kind"); kind != placementRuleKind {
+		kind, _, _ := unstructured.NestedString(manifest, "kind")
+		if kind != placementRuleKind && kind != placementKind {
 			continue
 		}
 
+		if kind == placementRuleKind && (!isPlR || p.plrCount.plr == 0) {
```
Comment (mprahl) [inline internal/plugin.go]:
```suggestion
		if kind == placementRuleKind {
```

File: internal/plugin.go [hunk 5]
```diff
@@ -424,10 +512,38 @@ func (p *Plugin) getPlrFromPath(plrPath string) (string, map[string]interface{},
...
+				)
+			}
+			if p.plrCount.plr == 0 {
+				return "", nil, fmt.Errorf(
+					"the placement %s specified a placementRule kind but should specify a placement kind in placement.placementPath",
+					plrPath,
+				)
+			}
+		}
+		if kind == placementKind && (isPlR || p.plrCount.pl == 0) {
```
Comment (mprahl) [inline internal/plugin.go]:
```suggestion

		if kind == placementKind {
```

File: internal/plugin.go [hunk 6]
```diff
@@ -415,7 +503,7 @@ func (p *Plugin) createPolicy(policyConf *types.PolicyConfig) error {
 // getPlrFromPath finds the placement rule manifest in the input manifest file. It will return
 // the name of the placement rule, the unmarshaled placement rule manifest, and an error. An error
 // is returned if the placement rule manifest cannot be found or is invalid.
-func (p *Plugin) getPlrFromPath(plrPath string) (string, map[string]interface{}, error) {
+func (p *Plugin) getPlrFromPath(plrPath string, isPlR bool) (string, map[string]interface{}, error) {
```
Comment (mprahl) [inline internal/plugin.go]: Could you please document `isPlR` in the method comment?

File: internal/plugin.go [hunk 7]
```diff
@@ -424,10 +512,38 @@ func (p *Plugin) getPlrFromPath(plrPath string) (string, map[string]interface{},
...
 			continue
 		}
 
+		if kind == placementRuleKind && (!isPlR || p.plrCount.plr == 0) {
+			if !isPlR {
+				return "", nil, fmt.Errorf(
+					"the placement %s specified a placementRule kind but expected a placement kind", plrPath,
+				)
+			}
+			if p.plrCount.plr == 0 {
```
Comment (mprahl) [inline internal/plugin.go]: Why is this check needed since you have `if !isPlR`? The caller of this method knows whether or not if this is supposed to be a `PlacementRule` or `Placement` and if it doesn't match what is in the manifest file, then it's always an error.

File: internal/plugin.go [hunk 8]
```diff
@@ -424,10 +512,38 @@ func (p *Plugin) getPlrFromPath(plrPath string) (string, map[string]interface{},
...
+				)
+			}
+		}
+		if kind == placementKind && (isPlR || p.plrCount.pl == 0) {
+			if isPlR {
+				return "", nil, fmt.Errorf(
+					"the placement %s specified a placement kind but expected a placementRule kind", plrPath,
+				)
+			}
+			if p.plrCount.pl == 0 {
```
Comment (mprahl) [inline internal/plugin.go]: Same comment as above.

**Review by mprahl — CHANGES_REQUESTED** [review]

**Comment by mprahl** [comment]
@dhaiducek I also think it'd be good to potentially rename the methods that reference `placementRule` to just be `placement`.

File: internal/plugin.go (hunk 7 above)
Comment (dhaiducek) [reply internal/plugin.go]: The first check is whether a PlacementRule or Placement is expected. This second check is to ensure only one type of placement is used across the generated policies.

Comment (dhaiducek) [reply internal/plugin.go]: I suppose I could make this clearer by instead of having the counter `plrCount` in the struct I could replace it with a `usingPlR` boolean and have the counts exist only in `assertValidConfig`.

Comment (mprahl) [reply internal/plugin.go]: I like this approach!

File: internal/plugin.go [hunk 9]
```diff
@@ -22,6 +22,8 @@ const (
 	placementBindingKind       = "PlacementBinding"
 	placementRuleAPIVersion    = "apps.open-cluster-management.io/v1"
 	placementRuleKind          = "PlacementRule"
+	placementAPIVersion        = "cluster.open-cluster-management.io/v1alpha1"
```
Comment (mprahl) [inline internal/plugin.go:25]: It feels a little odd that we are deprecating placement rules usage when placements are still in alpha based on the API version.

If placement rules have been announced as being deprecated then I agree with this approach and we should link to the announcement in the read me. Otherwise, I think we should just support both (not marking one as depreciated) until placement rules are announced as deprecated.

File: internal/plugin.go [hunk 10]
```diff
@@ -36,15 +38,18 @@ type Plugin struct {
...
 	Policies       []types.PolicyConfig `json:"policies" yaml:"policies"`
-	// A set of all placement rule names that have been processed or generated
-	allPlrs map[string]bool
+	// A set of all placement names that have been processed or generated
+	allPlcs map[string]bool
 	// This is a mapping of cluster selectors formatted as the return value of getCsKey to placement
-	// rule names. This is used to find common cluster selectors that can be consolidated to a
-	// single placement rule.
-	csToPlr      map[string]string
+	// names. This is used to find common cluster selectors that can be consolidated to a single
```
Comment (mprahl) [inline internal/plugin.go]:
```suggestion
	// names. This is used to find common cluster/label selectors that can be consolidated to a single
```

File: internal/plugin.go [hunk 11]
```diff
@@ -377,8 +440,41 @@ func (p *Plugin) assertValidConfig() error {
...
+	if plCount.plc != 0 && plCount.plr != 0 {
+		return fmt.Errorf(
+			"may not use a mix of Placement and PlacementRule for policies; found %d Placement and %d PlacementRule",
+			plCount.plc, plCount.plr,
+		)
+	}
+
+	if plCount.plr != 0 {
+		p.usingPlR = true
 	}
```
Comment (mprahl) [inline internal/plugin.go]: Optional:
```suggestion
	p.usingPlR = plCount.plr != 0
```

File: internal/plugin.go [hunk 12]
```diff
@@ -424,127 +520,176 @@ func (p *Plugin) createPolicy(policyConf *types.PolicyConfig) error {
...
-// getPlrFromPath finds the placement rule manifest in the input manifest file. It will return
-// the name of the placement rule, the unmarshaled placement rule manifest, and an error. An error
-// is returned if the placement rule manifest cannot be found or is invalid.
-func (p *Plugin) getPlrFromPath(plrPath string) (string, map[string]interface{}, error) {
-	manifests, err := unmarshalManifestFile(plrPath)
+// getPlcFromPath finds the placement manifest in the input manifest file. The isPlR boolean
+// signals whether the path should point to a PlacementRule kind (as opposed to a Placement kind).
+// It will return the name of the placement, the unmarshaled placement manifest, and an
+// error. An error is returned if the placement manifest cannot be found or is invalid.
+func (p *Plugin) getPlcFromPath(plcPath string, isPlR bool) (string, map[string]interface{}, error) {
```
Comment (mprahl) [inline internal/plugin.go]: Is the `isPlR` argument needed since after running the validation (i.e. `assertValidConfig`) you know that if `p.usingPlR` is set that only placement rules are expected and vice-versa?

**Review by mprahl — COMMENTED** [review]
This looks good! Nice job of thinking of all the edge cases.

I left a few comments. Let me know when this is ready for review again after you've added the tests.

File: internal/plugin.go (hunk 9 above)
Comment (dhaiducek) [reply internal/plugin.go:25]: Good point. I found a [thread in Slack](https://coreos.slack.com/archives/C02FF3579K9/p1636441525036700?thread_ts=1636375640.035500&cid=C02FF3579K9) stating PlacementRules are being deprecated in `2.5` and removed in `2.8`. But in the community they prefer GRC to not rely on App Lifecycle (where PlacementRule lives) and I think are pushing Placement harder. I think upstream I'm comfortable stating its deprecation but you're right--if we pull these changes downstream we should wait until we can link to an announcement before saying it's deprecated.

File: internal/plugin.go (hunk 11 above)
Comment (dhaiducek) [reply internal/plugin.go]: Nifty one-liner--I'll take it!

File: internal/plugin.go (hunk 12 above)
Comment (dhaiducek) [reply internal/plugin.go]: Yes because a user could use `placementRulePath` but specify a Placement kind there. `assertValidConfig` wouldn't be able to verify that. But that's a good point. The use of `placementRulePath` vs `placementPath` could be handled before entering the manifest retrieval. I'll update the code.

Comment (dhaiducek) [reply internal/plugin.go]: I thought about it more, and I see your point now. Yes, I'll scrap `isPlR`.

File: internal/plugin.go (hunk 9 above)
Comment (dhaiducek) [reply internal/plugin.go:25]: ~Shoot. There actually is a `v1`. I was going off old references.~
Scratch that. `Placement` is only `v1alpha1`(along with the ClusterPool kinds).

Comment (dhaiducek) [reply internal/plugin.go:25]: Okay, I guess I made some assumptions and hadn't done all of my homework. It looks like `Placement` has additional dependencies on other `v1alpha1` APIs like `ManagedClusterSet`. I may have to consider adding additional logic to create a `ManagedClusterSetBinding`... 🤔

Comment (dhaiducek) [reply internal/plugin.go:25]: Okay, I've removed statements about deprecation but left Placement as the default. Placement does appear to assume a `ManagedClusterSet` has been created also, but I think management of cluster sets is outside the realm of the generator.

File: internal/plugin.go [hunk 13]
```diff
@@ -91,41 +96,41 @@ func (p *Plugin) Generate() ([]byte, error) {
...
 		}
 
-		// If there is more than one policy associated with a placement rule but no default binding name
+		// If there is more than one policy associated with a placement but no default binding name
 		// specified, throw an error
 		if len(policyConfs) > 1 && p.PlacementBindingDefaults.Name == "" {
 			return nil, fmt.Errorf(
-				"placementBindingDefaults.name must be set but is empty (mutiple policies were found for the PlacementBinding to placement '%s')",
-				plrName,
+				"placementBindingDefaults.name must be set but is empty (multiple policies were found for the PlacementBinding to placement %s)",
```
Comment (mprahl) [inline internal/plugin.go:132]: You have quite the eagle eye catching that typo!

File: internal/plugin.go [hunk 14]
```diff
@@ -326,48 +343,94 @@ func (p *Plugin) assertValidConfig() error {
...
 
+	// Validate default Placement settings
+	if p.PolicyDefaults.Placement.PlacementRulePath != "" && p.PolicyDefaults.Placement.PlacementPath != "" {
+		return errors.New(
+			"policyDefaults must provide only one of placement.placementPath or placement.placementRulePath",
+		)
+	}
+	if len(p.PolicyDefaults.Placement.ClusterSelectors) > 0 && len(p.PolicyDefaults.Placement.LabelSelector) > 0 {
+		return errors.New(
+			"policyDefaults must provide only one of placement.labelSelector or placement.clusterselectors",
```
Comment (mprahl) [inline internal/plugin.go]:
```suggestion
			"policyDefaults must provide only one of placement.labelSelector or placement.clusterSelectors",
```

**Review by mprahl — APPROVED** [review]
Great job!

/hold to fix the typo.

File: internal/plugin.go (hunk 13 above)
Comment (dhaiducek) [reply internal/plugin.go:132]: Haha, actually I added a spell checker to VSCode. 🙂 It adds a fair bit of noise, but I'm kind of a stickler for spelling mistakes.

**Review by mprahl — APPROVED** [review]

**Comment by dhaiducek** [comment]
@mprahl I'm assuming this can be unheld now?
